Add note about named pipes on Windows in CLI help and fail on Windows when named pipe has incorrect form

piotr-iohk opened the pull request:
# Issue Number

#2340 

# Overview

- d597250c57d8fba12c3152889449c500f99fa4d0
  Add note about named pipes on Windows in CLI help for --node-socket
  
- 5437fcdeb7352ebb70f1229b428a59e86e660c20
  Add socket/pipe name validation to --node-socket option
  If the socket/pipe name is invalid, it will say one of:

  option --node-socket: Invalid file path

  option --node-socket: Invalid pipe name. Windows named pipes are of the form \\.\pipe\cardano-node


# Comments

Named pipes are used for communication between cardano-node and cardano-wallet on Windows. As far as I see it wasn't documented anywhere.
Before:
```
cardano-wallet serve
...
  --node-socket FILE       Path to the node's domain socket.
...
```
After:
```
cardano-wallet serve
...
  --node-socket FILE       Path to the node's domain socket file (POSIX) or pipe
                           name (Windows). Note: Maximum length for POSIX socket
                           files is approx. 100 bytes. Note: Windows named pipes
                           are of the form \\.\pipe\cardano-node

...
```

Commit:
Add note about named pipes on Windows in CLI help for --node-socket

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs:
```diff
@@ -242,7 +242,9 @@ nodeSocketOption :: Parser FilePath
-    <> help "Path to the node's domain socket."
+    <> help "Path to the node's domain socket. \
+            \Note: named pipes are used on Windows \
+            \e.g. \\\\.\\pipe\\cardano-node"
```
Maybe enhance it something like this? 

```suggestion
    <> help "Path to the node's domain socket file (POSIX) or pipe name (Windows). \
            \ Note: Maximum length for POSIX socket files is approx. 100 bytes. \
            \ Note: Windows named pipes are of the form \\\\.\\pipe\\cardano-node"
```

rvl reviewed (approved):
Ah, of course! Nice one. I totally forgot about that.

We should update the wiki too.

There are socket filename gotchas on Linux and macOS too.

rvl committed:
Add socket/pipe name validation to --node-socket option

If the socket/pipe name is invalid, it will say one of:

  option --node-socket: Invalid file path

  option --node-socket: Invalid pipe name. Windows named pipes are of the form \\.\pipe\cardano-node

piotr-iohk commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/LaunchSpec.hs:
```diff
@@ -0,0 +1,39 @@
+            , ( "invalid",          "",             err, err )
+            , ( "valid pipename",   "\\\\.pipe\\a", ok,  ok  )
+            , ( "invalid pipename", "\\\\.pipe\\",  ok,  err )
+            , ( "windows path",     "c:\\windows",  ok,  err )
```
:heart:

KtorZ commented on lib/shelley/exe/cardano-wallet.hs:
```diff
@@ -223,6 +225,11 @@ cmdServe = command "serve" $ info (helper <*> helper' <*> cmd) $ mempty
+                let socket = T.pack nodeSocket
+                when (os == "mingw32" && not ("\\\\.\\pipe\\" `T.isPrefixOf` socket)) $ do
+                    logError tr (MsgInvalidWinNamedPipe socket)
+                    exitWith $ ExitFailure 34
```
This validation is now redundant with the validation done in the command-line I believe :thinking:

KtorZ reviewed (approved):
bors merge

rvl committed:
LaunchSpec: Fix windows valid pipe name test case
